bpaotu/bpaotu/views.py

```diff
@@ -593,3 +613,115 @@ def dev_only_ckan_check_permissions(request):
+
+        return cache.get(key)
+    except Exception as e:
+        logger.error(str(e))
+
+
+def _set_cached_item(key, value):
+    try:
+        cache = caches['image_results']
+        cache.set(key, value, CACHE_7DAYS)
+    except Exception as e:
+        logger.error(str(e))
+
+    return True
+
+
+def _create_img_lookup_table():
+    '''
+    Create a lookup table of all images in ckan.
+    '''
+    if _get_cached_item(LOOKUP_TABLE_KEY) is None:
+        remote = ckanapi.RemoteCKAN(settings.BPA_PROD_URL, apikey=settings.CKAN_API_KEY)
+
+        packages = remote.action.package_search(
+            fq='tags:site-images',
+            include_private=True,
+            rows=100000
+        )['results']
+
+        logger.info('There are {} image packages.'.format(len(packages)))
+
+        lookup_table = defaultdict(list)
+
+        for i in packages:
+            try:
+                coords = (i['latitude'], i['longitude'])
+                img_url = i['resources'][0]['url']
+
+                lookup_table[coords].append(img_url)
+
+            except Exception as e:
+                logger.error("Either latitude or longitude is missing for: {}".format(img_url))
+                logger.error("Missing parameter is: {}".format(str(e)))
+
+                continue
+
+        lookup_table = dict(lookup_table)
+        _set_cached_item(LOOKUP_TABLE_KEY, lookup_table)
+
+
+def process_img(request=None, lat=None, lng=None, index=None):
+    '''
+    Return specified cached image or fetch image from ckan and resize before caching and returning.
+    '''
+    _create_img_lookup_table()
+
+    lookup_table = _get_cached_item(LOOKUP_TABLE_KEY)
+
+    img_url = lookup_table[(lat, lng)][int(index)]
+    img_name, img_ext = (img_url.split('/')[-1:])[0].split(".")
+    img_filename = img_name + "." + img_ext
+
+    key = img_filename
+
+    if _get_cached_item(key) is None:
+        try:
+            r = requests.get(img_url, headers={'Authorization': settings.CKAN_API_KEY})
+
+            with Image.open(BytesIO(r.content)) as img_obj:
+                # img_obj.save(img_filename)  # Saves image to a file on disk
+
+                # Resizing an image while maintaining aspect ratio:
+                # https://stackoverflow.com/questions/24745857/python-pillow-how-to-scale-an-image/24745969
+
+                maxsize = (MAX_WIDTH, MAX_HEIGHT)
+                img_obj.thumbnail(maxsize, Image.ANTIALIAS)
+
+                with BytesIO() as img_buf:
+                    # 1. Save image to BytesIO stream
+                    img_obj.save(img_buf, format=IMG_EXTENSION_TABLE[img_ext][0])
+
+                    # 2. Cache the BytesIO stream
+                    _set_cached_item(key, img_buf.getvalue())
+
+                width, height = img_obj.size
+                logger.info("Width: {} Height: {}".format(width, height))
+
+        except Exception as e:
+            logger.error("Error processing image: {}.".format(str(e)))
```

Comment:
again, don't catch the base `Exception` class - specialise this as needed. we shouldn't expect to have essentially random, unknown errors here to deal with. worst case the request bombs out, but at least then we know and we can fix it

Reply:
Fixed.